packages/babel-helpers/src/helpers.js

```diff
@@ -615,9 +615,14 @@ helpers.possibleConstructorReturn = helper("7.0.0-beta.0")`
   import assertThisInitialized from "assertThisInitialized";
 
   export default function _possibleConstructorReturn(self, call) {
-    if (call && (typeof call === "object" || typeof call === "function")) {
-      return call;
+    if (call) {
+      if (typeof call === "object" || typeof call === "function") {
+        return call;
+      } else {
```

Comment:
This logic doesn't work if the constructor returns, for example, `0` or `null` (can you add a test for a falsy value?).

It should be something like
```js
if (call && (typeof call === "object" || typeof call === "function")) {
  return call;
} else if (call !== void 0) {
  throw new TypeError("...");
}
```

Reply:
Or
```js
if (call !== void 0) {
  if (typeof call === "object" || typeof call === "function") {
    return call;
  } else {
    // throw errors
  }
}
```

Reply:
My version is 3 bytes shorter when using terser :stuck_out_tongue: 

```
function f(o){if(o&&("object"==typeof o||"function"==typeof o))return o;if(void 0!==o)throw new TypeError("...")}
function g(o){if(void 0!==o){if("object"==typeof o||"function"==typeof o)return o;throw new TypeError("...")}}
```

Also, I think yours incorrectly allows returning `null`.

Reply:
Oh holly `null`.

Reply:
but use of 'typeof' is a bad idea and kill perf.

Reply:
What do you suggest to use instead?

Reply:
Well. 'typeof obj === 'object';' is almost ~50% slower than other 'typeof'. And 'typeof number' return NaN even if it's not an number - in some cases. 

There exist 2 popular utility libraries that has a optimized check for this.

Personally I would have used some kind of meta data and bitwise masks

Reply:
> There exist 2 popular utility libraries that has an optimized check for this.

I only found `is-object` (which still uses [`typeof === "object"`](https://github.com/inspect-js/is-object/blob/main/index.js)) and `lodash.isObject` (which also [uses `typeof` and checks if it's `"object"`](https://github.com/lodash/lodash/blob/master/isObject.js)).

> Personally I would have used some kind of meta data and bitwise masks

`o` can be any possible JS value present in the JS code of our users, we cannot use bitwise mask

Reply:
I forgot the history behind this, but AngularJS had some good checks in their earlier versions. 1x of their library. I think they checked against the constructor. Also Google had a library a few years back - forgot the name - with fast checks.